### ISSUE TITLE
Switch to secondary URL delayed by over 30 seconds

### ISSUE THREAD

**Issue by SgtHugoStiglitz** [klassm/andFHEM#1003]
Since proximately a week it looks like the automated switch to the secondary URL (used for solving the issue with routing external requests back into own network) is delayed by over 30 seconds.

I tried to change the "Connection Timeout" to 1S and the "Command execution retries" to 0 but still the switch to the secondary URL will not happen before 30 seconds have passed.

I also tested it with a LTE (so not connected to WLAN) connection by setting up a wrong address as URL but the correct Address as "secondary URL". In this case it also will take over 30 seconds to successfully send the command.

I am using **https** (might be important to know due to the second last patch changing some certificate stuff)

Summary

**Within WLAN:**
URL: https://External/fhem
Secondary URL: https://Internal/fhem
_Commands are delayed by over 30 seconds._

When I change the config to
URL: https://internal/fhem
_Commands are sent instantly._

**Using LTE:**
URL: https://external/fhem
_commands are sent instantly._

URL: https://wrong address/fhem
secondary URL: https://external/fhem
_commands are delayed by over 30 seconds._

Thank you for making this tool!

**Comment by klassm** [comment]
Hi,

well, I suppose the issue is more or less the timeout for the first connection attempt. Before using the secondary URL andFHEM will always try to connect to the primary server. If the server does not respond within time (could be about 30s) the secondary URL will be used.

This is difficult to fix. How should I decided whether the server is just slow or down?

Matthias

**Comment by SgtHugoStiglitz** [comment]
Hi,

thank you for the quick reply.

The weird thing is it was working fine until approximately 1-2 weeks ago. Before that the switch to the secondary URL was done within 1 second or less.

It just recently started to have those huge delays.

Could you try to set up a wrong URL as main, and the proper URL as secondary and see if you are also having that delay?

If there is nothing you can do (or nothing you can find), maybe it would be nice to let the user adjust the timeout for the first connection attempt.

I was just wondering as this was working fine for several month and just started recently having those huge delays.

**Comment by SgtHugoStiglitz** [comment]
I looked in the Logs and I could find this: 

> D/li.klass.fhem.service.intent.DeviceIntentService( 6843): li.klass.fhem.constants.Actions.DEVICE_WIDGET_TOGGLE
I/FHEMWEBConnection( 6843): executeTask command set LED_Wand_TV on
I/FHEMWEBConnection( 6843): accessing URL https://nix.com:1234/fhem?XHR=1&cmd=set+LED_Wand_TV+on
I/art     ( 6843): Explicit concurrent mark sweep GC freed 14784(1006KB) AllocSpace objects, 48(40MB) LOS objects, 14% free, 45MB/53MB, paused 802us total 88.191ms
I/art     ( 6843): Explicit concurrent mark sweep GC freed 10(400B) AllocSpace objects, 0(0B) LOS objects, 14% free, 45MB/53MB, paused 780us total 48.801ms
I/art     ( 6843): Explicit concurrent mark sweep GC freed 5(160B) AllocSpace objects, 0(0B) LOS objects, 14% free, 45MB/53MB, paused 666us total 48.234ms
I/art     ( 6843): Explicit concurrent mark sweep GC freed 3(96B) AllocSpace objects, 0(0B) LOS objects, 14% free, 45MB/53MB, paused 491us total 32.061ms
I/art     ( 6843): Explicit concurrent mark sweep GC freed 3(96B) AllocSpace objects, 0(0B) LOS objects, 14% free, 45MB/53MB, paused 492us total 31.523ms
I/art     ( 6843): Explicit concurrent mark sweep GC freed 3(96B) AllocSpace objects, 0(0B) LOS objects, 14% free, 45MB/53MB, paused 491us total 31.256ms
I/FHEMWEBConnection( 6843): cannot connect to host
**I/FHEMWEBConnection( 6843): java.net.SocketTimeoutException: failed to connect to nix.com/192.64.147.142 (port 1234) after 20000ms**
I/FHEMWEBConnection( 6843): 	at libcore.io.IoBridge.connectErrno(IoBridge.java:169)
I/FHEMWEBConnection( 6843): 	at libcore.io.IoBridge.connect(IoBridge.java:122)
I/FHEMWEBConnection( 6843): 	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:183)
I/FHEMWEBConnection( 6843): 	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:456)
I/FHEMWEBConnection( 6843): 	at java.net.Socket.connect(Socket.java:882)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.Platform.connectSocket(Platform.java:139)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.Connection.connect(Connection.java:1171)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:380)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:289)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:373)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:323)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:491)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.DelegatingHttpsURLConnection.getResponseCode(DelegatingHttpsURLConnection.java:105)
I/FHEMWEBConnection( 6843): 	at com.android.okhttp.internal.http.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:25)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.fhem.FHEMWEBConnection.executeRequest(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.fhem.FHEMWEBConnection.executeRequest(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.fhem.FHEMWEBConnection.executeCommand(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.CommandExecutionService.execute(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.CommandExecutionService.execute(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.CommandExecutionService.executeSafely(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.device.ToggleableService.toggleState(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.intent.DeviceIntentService.toggleIntent(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.intent.DeviceIntentService.handleIntent(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.intent.ConvenientIntentService.handleTaskInternal(Unknown Source)
I/FHEMWEBConnection( 6843): 	at li.klass.fhem.service.intent.ConvenientIntentService.onHandleIntent(Unknown Source)
I/FHEMWEBConnection( 6843): 	at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:65)
I/FHEMWEBConnection( 6843): 	at android.os.Handler.dispatchMessage(Handler.java:102)
I/FHEMWEBConnection( 6843): 	at android.os.Looper.loop(Looper.java:145)
I/FHEMWEBConnection( 6843): 	at android.os.HandlerThread.run(HandlerThread.java:61)
I/FHEMWEBConnection( 6843): retrying request for alternate URL
I/FHEMWEBConnection( 6843): accessing URL https://myalternateurl.com:1234/fhem?XHR=1&cmd=set+LED_Wand_TV+on
I/CommandExecutionService( 6843): execute() - executing command=set LED_Wand_TV on, try=0
I/AppWidgetUpdateService( 6843): onHandleIntent() - updating all widgets (received REDRAW_ALL_WIDGETS)



Unfortunately I don't have logs from 2 Weeks ago where the switch happened immediately when the main URL was not reachable.
Is it maybe related to some other updates changing the java environment on my phone? (Non-rooted Galaxy S5)

**Comment by klassm** [comment]
I just tried that one:
* Setting the URL to some that does not exist (port 8085 instead of 8084 or so) - one that will return a 404.
* Entering the valid URL as secondary URL

AndFHEM immediately switches to the alternative URL:

> 03-27 13:39:30.126 11216-6379/li.klass.fhem I/FHEMWEBConnection: cannot connect to host
> java.net.ConnectException: failed to connect to www.klass.li/84.164.101.100 (port 8085) after 20000ms: isConnected failed: EHOSTUNREACH (No route to host)
> 03-27 13:39:30.126 11216-6379/li.klass.fhem I/FHEMWEBConnection: retrying request for alternate URL

Trying your example (https://nix.com:1234) also works for me. The app switches to the alternative URL immediately.

Out of pure interest - what happens if you access the primary URL in chrome (or some browser) on your phone? How long does Chrome take to find out that the URL cannot be opened?

Matthias

**Comment by SgtHugoStiglitz** [comment]
Hi Matthias,

thanks again for your time!

I did some more testing. As you guessed right, it also takes a long time before my browser tells me it cannot reach the page. Unfortunately I don't know if this was the case before I noticed those delays.

I tried your invalid url "http://www.klass.li:8085 (I think I accidentally sent you that report, sorry)
Here the switch to alternate URL happens instantly.

But for example if i put "klaas.li:8085"  it has a long delay before it switches to the alternate URL.

I tried installing andfhem on my old Galax S2 and it seems to have similar Delays.

Do you have any .apk from a previous andFHEM setup which I could try?

Michael

**Comment by klassm** [comment]
Hm I'd have to build an old version - but I don't think this would be helpful in any case, as nothing as change here since january.

The root cause is your web server (i.e. that klaas.li URL). Usually it will provide some kind of response, even if it is only a 404 - not found. Whenever the server's response is pending over a long time, the client will wait for the response to come. The server may just take some time to generate the response (this might take some time, which is why andFHEM has set this option to 20s) or just never answert. There is no option to find that out.

In other words - you have to find out why your webserver does not respond on that URL. This might be a router problem or one with your ISP.

Matthias
